Add Trino 403 release notes (#14977)
trinodb/trino: docs/src/main/sphinx/release/release-403.md

# Release 403 (9 Nov 2022)

## General

> * Round `TimeWithTimeZone` values when `PARAMETRIC_DATETIME` is not set. ({issue}`14950`)

Reviewer: This should be described in terms of user facing behavior. In particular, this is about fixing handling of variable precision time with time zone types for legacy clients that only understand precision = 3.

> * Round `TimeWithTimeZone` values when `PARAMETRIC_DATETIME` is not set. ({issue}`14950`)

Reviewer: No, it's a client protocol flag that clients can provide

> * Round `TimeWithTimeZone` values when `PARAMETRIC_DATETIME` is not set. ({issue}`14950`)

Reviewer: See https://github.com/trinodb/trino/blob/master/client/trino-client/src/main/java/io/trino/client/ClientCapabilities.java#L26

* Fix query deadlock in multi-join queries where broadcast join size is
  underestimated. ({issue}`14948`)

> * Fix incorrect results when `min(x, n)` or `max(x, n)` is used as a window

Reviewer 2: do we need to detail what was wrong before?

Reviewer: I don't think we need to spell it out here. If readers want more details, they can navigate to the issue.

  function. ({issue}`14886`)
* Fix failure for certain queries involving joins over partitioned tables. ({issue}`14317`)
* Fix incorrect order of parameters in `DESCRIBE INPUT` when they appear in a
  `WITH` clause. ({issue}`14738 `)

## BigQuery connector

> * Improve performance of `SHOW SCHEMAS` by adding a dataset metadata cache. This

Reviewer: What's a "dataset"? That word seems redundant. I think just saying "adding a metadata cache" is sufficient.

>   can be configured with the `bigquery.dataset-cache-ttl` property, which has a

Reviewer 3: Change to `bigquery.metadata.cache-ttl`.

  default value of `0ms` (disabled). ({issue}`14729`)

## ClickHouse connector

> * Add support for ClickHouse `DateTime(timezone)` type in read-only operations. ({issue}`13541`)

Reviewer 2: empty line after

> ## Hive connector

Reviewer 3: A new line is missing.

* Don't show the unsupported `AUTHORIZATION ROLE` property in the result of

>   `SHOW CREATE SCHEMA` when the access control doesn't support roles. ({issue}`8817`)

Reviewer 2: maybe

```suggestion
  `SHOW CREATE SCHEMA` if the access control doesn't support roles. ({issue}`8817`)
```

## Iceberg connector

> * Improve performance and reduce storage costs of the `expire_snapshots` table

Reviewer 2: storage cost? do you mean a dollar value? Or storage size or allocation or whatever?

>   procedure on S3 storage. ({issue}`14434`)

Reviewer 2: only on s3 ?? also S3-compatible at least I assume...

> * Allow registering existing table files in the metastore. This can be done

Reviewer 2: ```suggestion
* Allow registering existing table files in the metastore with the new ... 
```

# Release 403 (9 Nov 2022)

## General

> * Fix handling of `TimeWithTimeZone` values by rounding when the client does not

Reviewer: TimeWithTimeZone is an internal implementation details. Rephrase as:

> Fix handling of `time(n) with time zone` when clients do not support variable precision temporal types.

# Release 403 (12 Nov 2022)

## General

* Add the amount of data read from external sources during table scans to
  `EXPLAIN ANALYZE`. ({issue}`14907`)
* Fix handling of `time(n) with time zone` when clients do not support variable

>   precision temporal types. ({issue}`14950`)

Reviewer 4: Too vague what the fix is (i.e. "Fix handling" is vague) - maybe `Round `time(n) with time zone` values to precision 3 when clients do not support variable precision temporal types.`?

Reviewer: The thing is, this is how it was supposed to work all along. When we introduced variable precision temporal types, clients only understood values with 3 digits of precision (and often failed with anything else depending on how they were set up to parse them). To maintain compatibility, we added a flag that presents the values to the client using at most three digits, but we missed it for time with time zone.
This change just corrects that mistake.

What we want to convey is not that the value gets rounded, but that clients can now handle (not fail) when querying time with time zone columns.

Maybe “fix potential failure when clients …”

Reviewer 4: "Fix potential failure" makes more sense indeed.

* Fix query deadlock in multi-join queries where broadcast join size is
  underestimated. ({issue}`14948`)
* Fix incorrect results when `min(x, n)` or `max(x, n)` is used as a window
  function. ({issue}`14886`)
* Fix failure for certain queries involving joins over partitioned tables. ({issue}`14317`)
* Fix incorrect order of parameters in `DESCRIBE INPUT` when they appear in a
  `WITH` clause. ({issue}`14738`)

## Security

* Allow access token passthrough when using OAuth 2.0 with refresh tokens
  enabled. ({issue}`14949`)

> ## Fault-tolerant execution

Reviewer: Move this to the general section. We don’t usually create dedicated sections for each feature (aside from higher level components like connectors, clients, web ui, or general areas like the SPI or security)

> * Override the values of the `exchange.compression-enabled`,

Reviewer 4: Override -> Automatically set/configure?

Author: I think override is more clear for the sake of establishing that even if you otherwise have configuration values for these properties, they're being ignored when you turn on FTE.

Reviewer: We’re going to have to fix that. Silently overriding the values is error prone. It also goes against the principles of how airlift configuration works - when there are unused properties or usage of defunct properties, the server should refuse to start. cc @arhimondr

Reviewer 4: It doesn't silently override. It's using `bindConfigDefaults` which only change default values - not any user specified values.

Reviewer: In that case, I don't think we even need to document this. If the user is already setting these explicitly, their settings will be preserved. If they are not, they get handled automatically.

Reviewer 5: Yeah, let's not document it. We should reflect these changes on the fault tolerant execution configuration guide though, so we no longer suggest users to set these properties explicitly.

Author: By "not document" do we mean not include a release note at all?

Reviewer: Correct

  `query.low-memory-killer.delay`, `query.remote-task.max-error-duration`, and
  `task.concurrency` configuration properties to the recommended values for
  fault-tolerant execution when it is enabled. ({issue}`14957`)

## BigQuery connector

* Improve performance of `SHOW SCHEMAS` by adding a metadata cache. This can be

>   configured with the `bigquery.metadata.cache-ttl` catalog property, which has
>   a default value of `0ms` (disabled). ({issue}`14729`)

Reviewer 4: ```suggestion
  configured with the `bigquery.metadata.cache-ttl` catalog property, which is
  disabled by default. ({issue}`14729`)
```

* Fix failure when a [row access 
  policy](https://cloud.google.com/bigquery/docs/row-level-security-intro)
  returns an empty result. ({issue}`14760`)

## ClickHouse connector

> * Add support for ClickHouse `DateTime(timezone)` type in read-only operations. ({issue}`13541`)

Reviewer 4: It'd be good to let authors know when RNs are editorialized. Historically we've always mentioned the data type which a type is being mapped to in release notes - specially since we don't link to docs from here.

In this case the original from https://github.com/trinodb/trino/issues/14878#issuecomment-1303857828 seems better.

Author: When a mapping changes, that may impact existing workflows that rely on the old mapping. When a mapping can be used for writes, you need to know what type to use in Trino to write to the connector. Those are when I've made sure to include both types in the RN. But when a read-only mapping is being added, users don't need to be strictly told what type it appears as in Trino, because it will be evident the first time they query a ClickHouse table with a `DateTime(timezone)` column, and they can otherwise look in the docs or at the linked PR as necessary. There's no worries about anything breaking, nor do users need guidance beyond what's stated here to use this new functionality. I'm still open to changing it for consistency's sake, but this was a conscious decision made to prioritize conciseness where further info didn't seem required.

As far as communicating editorialization goes, that's probably worth a discussion. I'll follow-up with you on the Trino Slack - there's a few different ways we can go about this.

Reviewer 2: As a general rule .. we editorialize all suggestions for the RN entry.If you want to confirm the ones for your merged PRs that would certainly be good, but we want to avoid pinging maintainers and wasting their valuable time with grammar edits and other things. Also keep in mind each RN is at minimum fully reviewed by @colebow @martint and myself

Reviewer 4: > But when a read-only mapping is being added, users don't need to be strictly told what type it appears as in Trino, because it will be evident the first time they query a ClickHouse table with a DateTime(timezone) column, and they can otherwise look in the docs or at the linked PR as necessary.

While this is pragamatic it's not practically useful. Since you can CTAS/INSERT INTO SELECT which would get affected by the read mapping as well. It's useful information to have.

I'd not ask for it to be added if we linked to the type mapping docs from here but we don't. Let's not make things harder for users by having to manually search or know what the correct page is for the information they are seeking.

## Hive connector

* Don't show the unsupported `AUTHORIZATION ROLE` property in the result of
  `SHOW CREATE SCHEMA` if the access control doesn't support roles. ({issue}`8817`)

## Iceberg connector

> * Improve performance and reduce storage usage of the `expire_snapshots` table

Reviewer 4: "reduce storage usage" - what does it mean?

Or did you mean:

```suggestion
* Improve performance and reduce storage used by `expire_snapshots` table
```

  procedure on S3-compatible storage. ({issue}`14434`)
* Allow registering existing table files in the metastore with the new

>   `register_table` procedure. ({issue}`13552`)

Reviewer 4: Do we have docs? If so we might want to link to them from here.

> # Release 403 (15 Nov 2022)

Reviewer 2: are we doing tomorrow?

Author: If no, I'll update again.

## General

* Add the amount of data read from external sources during table scans to
  `EXPLAIN ANALYZE`. ({issue}`14907`)

> * Override the values of the `exchange.compression-enabled`,

Reviewer 2: Automatically override ..

  `query.low-memory-killer.delay`, `query.remote-task.max-error-duration`, and
  `task.concurrency` configuration properties to the recommended values for
  fault-tolerant execution when it is enabled. ({issue}`14957`)
* Improve performance of [exchange management](fte-exchange-manager) encryption

>   in fault-tolerant execution. ({issue}`14941`)

Author: @arhimondr - look good to you?

* Fix potential failure when clients do not support variable precision temporal
  types. ({issue}`14950`)
* Fix query deadlock in multi-join queries where broadcast join size is
  underestimated. ({issue}`14948`)
* Fix incorrect results when `min(x, n)` or `max(x, n)` is used as a window
  function. ({issue}`14886`)
* Fix failure for certain queries involving joins over partitioned tables. ({issue}`14317`)
* Fix incorrect order of parameters in `DESCRIBE INPUT` when they appear in a
  `WITH` clause. ({issue}`14738`)

## Security

> * Allow access token passthrough when using OAuth 2.0 with refresh tokens

Reviewer 2: ```suggestion
* Allow access token passthrough when using OAuth 2.0 authentication with refresh tokens
```

  enabled. ({issue}`14949`)

## BigQuery connector

* Improve performance of `SHOW SCHEMAS` by adding a metadata cache. This can be
  configured with the `bigquery.metadata.cache-ttl` catalog property, which is
  disabled by default. ({issue}`14729`)
* Fix failure when a [row access 
  policy](https://cloud.google.com/bigquery/docs/row-level-security-intro)
  returns an empty result. ({issue}`14760`)

## ClickHouse connector

* Add support for ClickHouse `DateTime(timezone)` type in read-only operations. ({issue}`13541`)

## Hive connector

> * Don't show the unsupported `AUTHORIZATION ROLE` property in the result of

Reviewer 2: Avoid showing the unsupported

>   `SHOW CREATE SCHEMA` if the access control doesn't support roles. ({issue}`8817`)

Reviewer 2: when the access control

## Iceberg connector

> * Improve performance and reduce storage used by the `expire_snapshots` table

Reviewer 2: "storage used" sounds kinda bad..  can we find something better?

  procedure on S3-compatible storage. ({issue}`14434`)
* Allow registering existing table files in the metastore with the new
  [`register_table` procedure](iceberg-register-table). ({issue}`13552`)

## MongoDB connector

* Add support for {doc}`/sql/delete`. ({issue}`14864`)
* Fix incorrect results for predicate pushdown applied to columns with `varchar`
  and `char` types when MongoDB collections have a collation. ({issue}`14900`)

## SQL Server connector

> * Fix incorrect results when non-transactional `INSERT` is disabled and bulk

Reviewer 2: results of what ... the INSERT query or something else ?? Did it use to fail entirely or insert wrong data or something else?

Reviewer 3: > results of what ... the INSERT query or something else ?? 

Results of `INSERT`

> Did it use to fail entirely or insert wrong data or something else?

Inserted wrong data.

Author: When bulk insert is enabled, the copy of the table that Trino loads ends up being empty. I think this is probably the most succinct way of explaining it without getting too into the weeds?

Reviewer 3: Please note that the wrong data were all null, not empty.

  underestimated. ({issue}`14948`)
* Fix incorrect results when `min(x, n)` or `max(x, n)` is used as a window
  function. ({issue}`14886`)
* Fix failure for certain queries involving joins over partitioned tables. ({issue}`14317`)
* Fix incorrect order of parameters in `DESCRIBE INPUT` when they appear in a
  `WITH` clause. ({issue}`14738`)
* Fix failure for queries involving `BETWEEN` predicates over `varchar` columns
  that contain temporal data. ({issue}`14954`)

## Security

* Allow access token passthrough when using OAuth 2.0 authentication with
  refresh tokens enabled. ({issue}`14949`)

## BigQuery connector

* Improve performance of `SHOW SCHEMAS` by adding a metadata cache. This can be
  configured with the `bigquery.metadata.cache-ttl` catalog property, which is
  disabled by default. ({issue}`14729`)
* Fix failure when a [row access policy](https://cloud.google.com/bigquery/docs/row-level-security-intro)
  returns an empty result. ({issue}`14760`)

## ClickHouse connector

* Add mapping for the ClickHouse `DateTime(timezone)` type to the Trino
  `timestamp(0) with time zone` type for read-only operations. ({issue}`13541`)

## Delta Lake connector

> * Fix statistics for `DATE` columns. ({issue}`15005`)

Reviewer 6: We should remove this.  https://github.com/trinodb/trino/pull/15005 has no practical end-user implications that I am aware of, unless "bechmarks" is a "user".

cc @raunaqmorarka

Reviewer 2: Well .. isn't this a performance improvement? If so .. there is user facing impact.

Reviewer 6: Not in any practical situation, so no, it's not user visible.

Reviewer 2: Okay .. then I am not understanding this. Whats the point of having these stats then.. or fixing them. If there is no impact.

Reviewer 7: The change is about fixing stats for DATE in a particular range (1900-01-02 to 1901-12-31). This range occurs in TPCDS and not having accurate stats results in bad plans for certain TPCDS queries. 
I have no idea how to assess if this is possible in user data, however we know that ppl run TPC benchmarks and care about it running faster.

Reviewer 6: in any case, "Fix statistics for `DATE` columns" is bad wording, since nothing was broken and nothing got fixed.

you can change this to 

"Add support for date column min/max statistics for values in range 1900-01-02 between 1901-12-31. Such dates occur in TPC-DS data sets."

This would be fathful, but also would tell the truth that real users don't care about this change, because no production workload will benefit. No-one runs TPC-DS as "production workloads".

Reviewer 7: My suggested RN in the PR was `Improve statistics for DATE columns in delta lake`.
It's inaccurate to claim that no production workload can ever benefit from it. How do you know that there isn't some dimension table in "real" data which contains birth dates of ppl from that time ?

Reviewer 6: > It's inaccurate to claim that no production workload can ever benefit from it. 

Lack of release notes entry doesn't claim that.

Existence of release notes entry claims that there is a benefit.

> How do you know that there isn't some dimension table in "real" data which contains birth dates of ppl from that time ?

How do you know there is?

Reviewer: I agree with the rewording to make what was fixed more explicit. We should reserve judgement on whether such data appears in production workloads. That is very situational. There are many domains that deal with data in that range (and older) - historical weather record, etc.

Reviewer 6: For applications dealing with real historical data the fix won't really help, as it moves the min supported min value by just a year.

# Release 403 (15 Nov 2022)

## General

* Add the amount of data read from external sources during table scans to
  `EXPLAIN ANALYZE`. ({issue}`14907`)
* Improve performance of worker-to-worker data transfer encryption when
  fault-tolerant execution is enabled. ({issue}`14941`)
* Improve performance of aggregations when input data does not contain nulls. ({issue}`14567`)
* Fix potential failure when clients do not support variable precision temporal
  types. ({issue}`14950`)
* Fix query deadlock in multi-join queries where broadcast join size is
  underestimated. ({issue}`14948`)
* Fix incorrect results when `min(x, n)` or `max(x, n)` is used as a window
  function. ({issue}`14886`)
* Fix failure for certain queries involving joins over partitioned tables. ({issue}`14317`)
* Fix incorrect order of parameters in `DESCRIBE INPUT` when they appear in a
  `WITH` clause. ({issue}`14738`)

> * Fix failure for queries involving BETWEEN predicates over `varchar` columns

Reviewer: Code quotes for `BETWEEN`

  that contain temporal data. ({issue}`14954`)

## Security

* Allow access token passthrough when using OAuth 2.0 authentication with
  refresh tokens enabled. ({issue}`14949`)

## BigQuery connector

* Improve performance of `SHOW SCHEMAS` by adding a metadata cache. This can be
  configured with the `bigquery.metadata.cache-ttl` catalog property, which is
  disabled by default. ({issue}`14729`)

> * Fix failure when a [row access 

Reviewer: Don't break label across lines. It's ok it the line extends beyond the soft margin

  policy](https://cloud.google.com/bigquery/docs/row-level-security-intro)
  returns an empty result. ({issue}`14760`)

## ClickHouse connector

* Add mapping for the ClickHouse `DateTime(timezone)` type to the Trino

>   `TIMESTAMP(0) WITH TIME ZONE` type for read-only operations. ({issue}`13541`)

Reviewer: Lower case `timestamp(0) with time zone`

# Release 403 (15 Nov 2022)

## General

> * Add the amount of data read from external sources during table scans to

Author: This would be inconsistent with every other RN we've had about adding information to explain analyze, which always do start with "Add"

> * Add the amount of data read from external sources during table scans to

Reviewer: True, however the other entries are a lot terser, so the consistency is already lost. E.g.,

> * Add output buffer utilization distribution to EXPLAIN ANALYZE VERBOSE
> * Add operator blocked time distribution to EXPLAIN ANALYZE VERBOSE

If we wanted consistency (not a goal, IMO), we might want to rephrase as:

> * Add amount of data read from external sources to `EXPLAIN ANALYZE`
